tf/tfprocess.py

```diff
@@ -523,134 +547,113 @@ def calculate_test_summaries(self, test_batches, steps):
-                    # Convolution weights need a transpose
-                    #
-                    # TF (kYXInputOutput)
-                    # [filter_height, filter_width, in_channels, out_channels]
-                    #
-                    # Leela/cuDNN/Caffe (kOutputInputYX)
-                    # [output, input, filter_size, filter_size]
-                    work_weights = tf.transpose(weights, [3, 2, 0, 1])
-                elif weights.shape.ndims == 2:
-                    # Fully connected layers are [in, out] in TF
-                    #
-                    # [out, in] in Leela
-                    #
-                    work_weights = tf.transpose(weights, [1, 0])
-                else:
-                    # Biases, batchnorm etc
-                    work_weights = weights
-                all_evals.append(work_weights)
-            self.pb_save_op = all_evals
-        nparrays = self.session.run(self.pb_save_op)
-        for e, nparray in enumerate(nparrays):
+        last_was_gamma = False
+        for weights in self.model.weights:
+            work_weights = None
+            if weights.shape.ndims == 4:
+                # Convolution weights need a transpose
+                #
+                # TF (kYXInputOutput)
+                # [filter_height, filter_width, in_channels, out_channels]
+                #
+                # Leela/cuDNN/Caffe (kOutputInputYX)
+                # [output, input, filter_size, filter_size]
+                work_weights = tf.transpose(a=weights, perm=[3, 2, 0, 1])
+            elif weights.shape.ndims == 2:
+                # Fully connected layers are [in, out] in TF
+                #
+                # [out, in] in Leela
+                #
+                work_weights = tf.transpose(a=weights, perm=[1, 0])
+            else:
+                # batch renorm has extra weights, but we don't know what to do with them.
+                if 'renorm' in weights.name:
+                    continue
+                # renorm has variance, but it is not the primary source of truth
+                if 'variance:' in weights.name and self.renorm_enabled:
+                    continue
+                # Renorm has moving stddev not variance, undo the transform to make it compatible.
+                if 'stddev:' in weights.name:
+                    all_tensors.append(tf.math.square(weights) - 1e-5)
+                    continue
+                # Biases, batchnorm etc
+                # pb expects every batch norm to have gammas, but not all of our
+                # batch norms have gammas, so manually add pretend gammas.
+                if 'beta:' in weights.name and not last_was_gamma:
+                    all_tensors.append(tf.ones_like(weights))
+                work_weights = weights.read_value()
+            all_tensors.append(work_weights)
+            last_was_gamma = 'gamma:' in weights.name
+
+        # HACK: model weights ordering is some kind of breadth first traversal,
```

Comment:
This looks too fragile. Should probably rework the `net.py` to assign the weights in proto based on the tensor names instead. I'll try to look into if that looks doable.

Reply:
I agree, but it was a lot easier to do this way than to do properly, so I'm going to stick with it for right now. If you want to give doing it right a shot, I'm happy :)